### PR TITLE
Partially fixed the bug UTF-8 characters (#715)

### PR DESCRIPTION
**Pull request by byzheng**
Partially fixed the bug to show UTF-8 characters through adding encoding.

Still need to add encoding to L312 at `html_document.R`, but I am not sure how to fix it.

```r
config <- site_config(input_file, encoding = ???)
```

### CODE REVIEW

**Comment by jjallaire** [comment]
Thanks for this. I'm traveling next week but should be able to merge this after that. Solving the problem in html_document may require changes in other parts of the package but should be do-able.

File: R/util.R [hunk 1]
```diff
@@ -107,7 +107,7 @@ file_name_without_shell_chars <- function(file) {
 as_tmpfile <- function(str) {
   if (length(str) > 0) {
     str_tmpfile <- tempfile("rmarkdown-str", fileext = ".html")
-    writeLines(str, str_tmpfile)
+    writeLines(str, str_tmpfile, useBytes =  TRUE)
```
Comment (yihui) [inline R/util.R]: The `str` probably needs to be encoded to UTF-8 first if `useBytes = TRUE`, i.e. `enc2utf8(str)`

File: R/html_document.R [hunk 2]
```diff
@@ -292,7 +292,7 @@ html_document <- function(toc = FALSE,
   # pre-processor for arguments that may depend on the name of the
   # the input file AND which need to inject html dependencies
   # (otherwise we could just call the pre_processor)
-  post_knit <- function(metadata, input_file, runtime, ...) {
+  post_knit <- function(metadata, input_file, runtime, encoding, ...) {
```
Comment (jjallaire) [inline R/html_document.R]: Note that render needs to be updated to pass the encoding (and to check whether post_knit accepts an encoding argument first).

**Comment by lcolladotor** [comment]
Hi,

I found this thread when searching `enc2utf8` in the `rmarkdown` repo. The [latest build report](http://bioconductor.org/checkResults/devel/bioc-LATEST/recount/malbec2-buildsrc.html) for my package `recount` at Bioconductor shows the following error:

```
Quitting from lines 354-375 (recount-quickstart.Rmd) 
Error: processing vignette 'recount-quickstart.Rmd' failed with diagnostics:
argumemt is not a character vector
Execution halted
```

The traceback shows that the error is at the `enc2utf8(text)` and was wondering if this is something you've seen. The full log is below; note that I'm using https://github.com/Bioconductor/BiocStyle's html_document2 there.

I expected that it would be easy to find the issue _argument_ is misspelled. But it doesn't show up when searching the `rmarkdown` or `BiocStyle` code.

Best,
Leo

```R
> library('recount')
> packageVersion('recount')
[1] ‘1.1.8’
> project_info <- abstract_search('GSE32465')
> download_study(project_info$project)
2016-12-02 13:22:59 downloading file rse_gene.Rdata to SRP009615
trying URL 'http://duffel.rail.bio/recount/SRP009615/rse_gene.Rdata'
Content type 'application/octet-stream' length 1383643 bytes (1.3 MB)
==================================================
downloaded 1.3 MB

> load(file.path(project_info$project, 'rse_gene.Rdata'))
>
> ## Extract the sample characteristics
> geochar <- lapply(split(colData(rse_gene), seq_len(nrow(colData(rse_gene)))), geo_characteristics)
> 
> ## Note that the information for this study is a little inconsistent, so we
> ## have to fix it.
> geochar <- do.call(rbind, lapply(geochar, function(x) {
+     if('cells' %in% colnames(x)) {
+         colnames(x)[colnames(x) == 'cells'] <- 'cell.line'
+         return(x)
+     } else {
+         return(x)
+     }
+ }))
> 
> ## We can now define some sample information to use
> sample_info <- data.frame(
+     run = colData(rse_gene)$run,
+     group = ifelse(grepl('uninduced', colData(rse_gene)$title), 'uninduced', 'induced'),
+     gene_target = sapply(colData(rse_gene)$title, function(x) { strsplit(strsplit(x,
+         'targeting ')[[1]][2], ' gene')[[1]][1] }),
+     cell.line = geochar$cell.line
+ )
> rse <- scale_counts(rse_gene)
> colData(rse)$group <- sample_info$group
> colData(rse)$gene_target <- sample_info$gene_target
> library('DESeq2')
> dds <- DESeqDataSet(rse, ~ gene_target + group)
converting counts to integer mode
> dds <- DESeq(dds, test = 'LRT', reduced = ~ gene_target, fitType = 'local')
estimating size factors
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
> res <- results(dds)
> plotMA(res, main="DESeq2 results for SRP009615")
> library('regionReport')
> ## Make it so that the report will be available as a vignette
> original <- readLines(system.file('DESeq2Exploration', 'DESeq2Exploration.Rmd',
+     package = 'regionReport'))
> vignetteInfo <- c(
+     'vignette: >',
+     '  %\\VignetteEngine{knitr::rmarkdown}',
+     '  %\\VignetteIndexEntry{Basic DESeq2 results exploration}',
+     '  %\\VignetteEncoding{UTF-8}'
+ )
> new <- c(original[1:16], vignetteInfo, original[17:length(original)])
> writeLines(new, 'SRP009615-results-template.Rmd')
> ## Now create the report
> report <- DESeq2Report(dds, res = res, project = 'SRP009615',
+     intgroup = c('group', 'gene_target'), outdir = '.',
+     output = 'SRP009615-results', device = 'png', template = 'SRP009615-results-template.Rmd', nBest = 10, nBestFeatures = 2)
Writing 9 Bibtex entries ... OK
Results written to file './SRP009615-results.bib'


processing file: SRP009615-results.Rmd
  |.                                                                  |   2%
  |...                                                                |   4%
  |....                                                               |   6%
  |......                                                             |   9%
label: setup (with options) 
List of 1
 $ bootstrap.show.message: logi FALSE

  |.......                                                            |  11%
  |.........                                                          |  13%
label: PCA
  |..........                                                         |  15%
  |...........                                                        |  17%
label: sampleDist
  |.............                                                      |  19%
  |..............                                                     |  21%
label: MAplotalpha
  |................                                                   |  23%
  |.................                                                  |  26%
label: MAplotalphaHalf
  |...................                                                |  28%
  |....................                                               |  30%
label: MAplotalpha-nBest
  |.....................                                              |  32%
  |.......................                                            |  34%
label: pvalueHistogram
  |........................                                           |  36%
  |..........................                                         |  38%
  |...........................                                        |  40%
  |.............................                                      |  43%
  |..............................                                     |  45%
  |...............................                                    |  47%
label: padjHistogram
  |.................................                                  |  49%
  |..................................                                 |  51%
  |....................................                               |  53%
  |.....................................                              |  55%
  |......................................                             |  57%
  |........................................                           |  60%
  |.........................................                          |  62%
  |...........................................                        |  64%
label: plotCounts
  |............................................                       |  66%
   inline R code fragments

  |..............................................                     |  68%
  |...............................................                    |  70%
  |................................................                   |  72%
  |..................................................                 |  74%
  |...................................................                |  77%
  |.....................................................              |  79%
  |......................................................             |  81%
  |........................................................           |  83%
  |.........................................................          |  85%
  |..........................................................         |  87%
  |............................................................       |  89%
  |.............................................................      |  91%
  |...............................................................    |  94%
  |................................................................   |  96%
  |.................................................................. |  98%
label: bibliography (with options) 
List of 3
 $ results: chr "asis"
 $ echo   : logi FALSE
 $ warning: logi FALSE

  |...................................................................| 100%
  ordinary text without R code


output file: SRP009615-results.knit.md

/usr/local/bin/pandoc +RTS -K512m -RTS SRP009615-results.utf8.md --to html --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output SRP009615-results.html --smart --email-obfuscation none --self-contained --standalone --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_collapsed=1 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /var/folders/cx/n9s558kx6fb7jf5z_pgszgb80000gn/T//Rtmpk2g6ps/BiocStyle/template.html --no-highlight --variable highlightjs=1 --number-sections --css /Library/Frameworks/R.framework/Versions/3.4/Resources/library/BiocStyle/resources/html/bioconductor2.css --variable 'theme:bootstrap' --include-in-header /var/folders/cx/n9s558kx6fb7jf5z_pgszgb80000gn/T//Rtmpk2g6ps/rmarkdown-strddf02c497d52.html --mathjax --variable 'mathjax-url:https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --variable code_folding=hide --variable code_menu=1 
Error in enc2utf8(text) : argumemt is not a character vector
In addition: Warning message:
In c.bibentry(knitcitations = citation("knitcitations"), regionReport = citation("regionReport")[1],  :
  method is only applicable to ‘bibentry’ objects
>     
> traceback()
9: writeLines(enc2utf8(text), ..., useBytes = TRUE)
8: writeUTF8(x, output)
7: overlay(metadata, input_file, output_file, ...)
6: post(metadata, input, output, clean, verbose)
5: output_format$post_processor(yaml_front_matter, utf8_input, output_file, 
       clean, !quiet)
4: render(paste0(output, ".Rmd"), output_format, clean = .advanced_argument("clean", 
       TRUE, ...))
3: eval(expr, envir = parent.frame())
2: with_wd(outdir, {
       file.copy(template, to = paste0(output, ".Rmd"))
       output_format <- .advanced_argument("output_format", "BiocStyle::html_document2", 
           ...)
       outputIsHTML <- output_format %in% c("html_document", "knitrBootstrap::bootstrap_document", 
           "BiocStyle::html_document2")
       if (!outputIsHTML) {
           if (device == "png") 
               warning("You might want to switch the 'device' argument from 'png' to 'pdf' for better quality plots.")
       }
       knitrBootstrapFlag <- packageVersion("knitrBootstrap") < 
           "1.0.0"
       if (knitrBootstrapFlag & output_format == "knitrBootstrap::bootstrap_document") {
           tmp <- knit_bootstrap(paste0(output, ".Rmd"), chooser = c("boot", 
               "code"), show_code = TRUE)
           res <- file.path(tmpdir, outdir, paste0(output, ".html"))
           unlink(paste0(output, ".md"))
       }
       else {
           res <- render(paste0(output, ".Rmd"), output_format, 
               clean = .advanced_argument("clean", TRUE, ...))
       }
       if (templateNull) 
           file.remove(paste0(output, ".Rmd"))
       if (browse) 
           browseURL(res)
   })
1: DESeq2Report(dds, res = res, project = "SRP009615", intgroup = c("group", 
       "gene_target"), outdir = ".", output = "SRP009615-results", 
       device = "png", template = "SRP009615-results-template.Rmd", 
       nBest = 10, nBestFeatures = 2)
> options(width = 120)
> devtools::session_info()
Session info -----------------------------------------------------------------------------------------------------------
 setting  value                                             
 version  R Under development (unstable) (2016-10-26 r71594)
 system   x86_64, darwin13.4.0                              
 ui       AQUA                                              
 language (EN)                                              
 collate  en_US.UTF-8                                       
 tz       America/New_York                                  
 date     2016-12-02                                        

Packages ---------------------------------------------------------------------------------------------------------------
 package              * version  date       source                                   
 acepack                1.4.1    2016-10-29 CRAN (R 3.4.0)                           
 annotate               1.53.0   2016-10-23 Bioconductor                             
 AnnotationDbi          1.37.0   2016-10-26 Bioconductor                             
 assertthat             0.1      2013-12-06 CRAN (R 3.4.0)                           
 backports              1.0.4    2016-10-24 CRAN (R 3.4.0)                           
 bibtex                 0.4.0    2014-12-31 CRAN (R 3.4.0)                           
 Biobase              * 2.35.0   2016-10-23 Bioconductor                             
 BiocGenerics         * 0.21.1   2016-12-01 Bioconductor                             
 BiocParallel           1.9.2    2016-11-18 Bioconductor                             
 BiocStyle            * 2.3.16   2016-12-01 Bioconductor                             
 biomaRt                2.31.1   2016-10-23 Bioconductor                             
 Biostrings             2.43.1   2016-11-17 Bioconductor                             
 bitops                 1.0-6    2013-08-17 CRAN (R 3.4.0)                           
 bookdown               0.3      2016-11-28 CRAN (R 3.4.0)                           
 BSgenome               1.43.1   2016-11-11 Bioconductor                             
 bumphunter             1.15.0   2016-10-23 Bioconductor                             
 checkmate              1.8.2    2016-11-02 CRAN (R 3.4.0)                           
 cluster                2.0.5    2016-10-08 CRAN (R 3.4.0)                           
 codetools              0.2-15   2016-10-05 CRAN (R 3.4.0)                           
 colorspace             1.3-1    2016-11-18 CRAN (R 3.4.0)                           
 data.table             1.9.8    2016-11-25 CRAN (R 3.4.0)                           
 DBI                    0.5-1    2016-09-10 CRAN (R 3.4.0)                           
 DEFormats              1.3.0    2016-10-23 Bioconductor                             
 derfinder              1.9.5    2016-11-30 Bioconductor                             
 derfinderHelper        1.9.3    2016-11-29 Bioconductor                             
 DESeq2               * 1.15.17  2016-11-29 Bioconductor                             
 devtools             * 1.12.0   2016-06-24 CRAN (R 3.4.0)                           
 digest                 0.6.10   2016-08-02 CRAN (R 3.4.0)                           
 doRNG                  1.6      2014-03-07 CRAN (R 3.4.0)                           
 downloader             0.4      2015-07-09 CRAN (R 3.4.0)                           
 DT                   * 0.2      2016-08-09 CRAN (R 3.4.0)                           
 edgeR                  3.17.4   2016-11-29 Bioconductor                             
 evaluate               0.10     2016-10-11 CRAN (R 3.4.0)                           
 foreach                1.4.3    2015-10-13 CRAN (R 3.4.0)                           
 foreign                0.8-67   2016-09-13 CRAN (R 3.4.0)                           
 Formula                1.2-1    2015-04-07 CRAN (R 3.4.0)                           
 genefilter             1.57.0   2016-10-23 Bioconductor                             
 geneplotter            1.53.0   2016-10-23 Bioconductor                             
 GenomeInfoDb         * 1.11.6   2016-11-17 Bioconductor                             
 GenomicAlignments      1.11.3   2016-11-17 Bioconductor                             
 GenomicFeatures        1.27.3   2016-11-29 Bioconductor                             
 GenomicFiles           1.11.3   2016-11-29 Bioconductor                             
 GenomicRanges        * 1.27.13  2016-12-01 Bioconductor                             
 GEOquery               2.41.0   2016-10-25 Bioconductor                             
 ggplot2              * 2.2.0    2016-11-11 CRAN (R 3.4.0)                           
 gridExtra              2.2.1    2016-02-29 CRAN (R 3.4.0)                           
 gtable                 0.2.0    2016-02-26 CRAN (R 3.4.0)                           
 highr                  0.6      2016-05-09 CRAN (R 3.4.0)                           
 Hmisc                  4.0-0    2016-11-01 CRAN (R 3.4.0)                           
 htmlTable              1.7      2016-10-19 CRAN (R 3.4.0)                           
 htmltools              0.3.5    2016-03-21 CRAN (R 3.4.0)                           
 htmlwidgets            0.8      2016-11-09 CRAN (R 3.4.0)                           
 httr                   1.2.1    2016-07-03 CRAN (R 3.4.0)                           
 IRanges              * 2.9.12   2016-12-01 Bioconductor                             
 iterators              1.0.8    2015-10-13 CRAN (R 3.4.0)                           
 jsonlite               1.1      2016-09-14 CRAN (R 3.4.0)                           
 knitcitations          1.0.7    2015-10-28 CRAN (R 3.4.0)                           
 knitr                * 1.15.1   2016-11-22 CRAN (R 3.4.0)                           
 knitrBootstrap         1.0.0    2016-11-15 Github (jimhester/knitrBootstrap@cdaa4a9)
 labeling               0.3      2014-08-23 CRAN (R 3.4.0)                           
 lattice                0.20-34  2016-09-06 CRAN (R 3.4.0)                           
 latticeExtra           0.6-28   2016-02-09 CRAN (R 3.4.0)                           
 lazyeval               0.2.0    2016-06-12 CRAN (R 3.4.0)                           
 limma                  3.31.6   2016-11-29 Bioconductor                             
 locfit                 1.5-9.1  2013-04-20 CRAN (R 3.4.0)                           
 lubridate              1.6.0    2016-09-13 CRAN (R 3.4.0)                           
 magrittr               1.5      2014-11-22 CRAN (R 3.4.0)                           
 markdown               0.7.7    2015-04-22 CRAN (R 3.4.0)                           
 Matrix                 1.2-7.1  2016-09-01 CRAN (R 3.4.0)                           
 matrixStats            0.51.0   2016-10-09 CRAN (R 3.4.0)                           
 memoise                1.0.0    2016-01-29 CRAN (R 3.4.0)                           
 munsell                0.4.3    2016-02-13 CRAN (R 3.4.0)                           
 nnet                   7.3-12   2016-02-02 CRAN (R 3.4.0)                           
 pheatmap             * 1.0.8    2015-12-11 CRAN (R 3.4.0)                           
 pkgmaker               0.22     2014-05-14 CRAN (R 3.4.0)                           
 plyr                   1.8.4    2016-06-08 CRAN (R 3.4.0)                           
 qvalue                 2.7.0    2016-10-23 Bioconductor                             
 R6                     2.2.0    2016-10-05 CRAN (R 3.4.0)                           
 RColorBrewer         * 1.1-2    2014-12-07 CRAN (R 3.4.0)                           
 Rcpp                   0.12.8   2016-11-17 CRAN (R 3.4.0)                           
 RCurl                  1.95-4.8 2016-03-01 CRAN (R 3.4.0)                           
 recount              * 1.1.8    2016-11-29 Bioconductor                             
 RefManageR             0.13.1   2016-11-13 CRAN (R 3.4.0)                           
 regionReport         * 1.9.3    2016-11-29 Bioconductor                             
 registry               0.3      2015-07-08 CRAN (R 3.4.0)                           
 rentrez                1.0.4    2016-10-26 CRAN (R 3.4.0)                           
 reshape2               1.4.2    2016-10-22 CRAN (R 3.4.0)                           
 RJSONIO                1.3-0    2014-07-28 CRAN (R 3.4.0)                           
 rmarkdown              1.2      2016-11-21 CRAN (R 3.4.0)                           
 rngtools               1.2.4    2014-03-06 CRAN (R 3.4.0)                           
 rpart                  4.1-10   2015-06-29 CRAN (R 3.4.0)                           
 rprojroot              1.1      2016-10-29 CRAN (R 3.4.0)                           
 Rsamtools              1.27.3   2016-11-29 Bioconductor                             
 RSQLite                1.1      2016-11-27 CRAN (R 3.4.0)                           
 rstudioapi             0.6      2016-06-27 CRAN (R 3.4.0)                           
 rtracklayer            1.35.1   2016-10-29 Bioconductor                             
 S4Vectors            * 0.13.5   2016-12-01 Bioconductor                             
 scales                 0.4.1    2016-11-09 CRAN (R 3.4.0)                           
 stringi                1.1.2    2016-10-01 CRAN (R 3.4.0)                           
 stringr                1.1.0    2016-08-19 CRAN (R 3.4.0)                           
 SummarizedExperiment * 1.5.3    2016-11-11 Bioconductor                             
 survival               2.40-1   2016-10-30 CRAN (R 3.4.0)                           
 tibble                 1.2      2016-08-26 CRAN (R 3.4.0)                           
 VariantAnnotation      1.21.9   2016-12-01 Bioconductor                             
 withr                  1.0.2    2016-06-20 CRAN (R 3.4.0)                           
 XML                    3.98-1.5 2016-11-10 CRAN (R 3.4.0)                           
 xtable                 1.8-2    2016-02-05 CRAN (R 3.4.0)                           
 XVector                0.15.0   2016-10-23 Bioconductor                             
 yaml                   2.1.14   2016-11-12 CRAN (R 3.4.0)                           
 zlibbioc               1.21.0   2016-10-23 Bioconductor                             
> 
```

**Comment by yihui** [comment]
@lcolladotor I'd recommend you to report this to BiocStyle and its authors can report back to us if they found it was a problem on our side. We have had a headache with BiocStyle recently, and I'm not sure if this is another one. Actually you can easily tell if it is our problem or theirs by changing the BiocStyle format to an rmarkdown format like `html_document`.

**Comment by lcolladotor** [comment]
Thanks @yihui. It does seem to be a `BiocStyle` issue and I have reported it accordingly there: https://github.com/Bioconductor/BiocStyle/issues/20.

Best,
Leo

**Comment by aoles** [comment]
I confirm that there has been a problem in one of the *BiocStyle* functions, which I've fixed in https://github.com/Bioconductor/BiocStyle/commit/333a30e54111cd262ca4e8244cc34f521e4d4745. Funny thing, this bug was introduced just recently while removing dependencies on internal functions from **rstudio**'s packages, following the complains in https://github.com/Bioconductor/BiocStyle/issues/18

@yihui thank you for your patience and sorry again for bothering you with our Bioconductor-related issues.

**Pull request merged**